fiaas_deploy_daemon/extension_hook_caller.py

```diff
@@ -0,0 +1,68 @@
+#!/usr/bin/env python
+# -*- coding: utf-8
+
+# Copyright 2017-2021 The FIAAS Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import json
+import logging
+from collections import OrderedDict
+
+from six import string_types
+
+
+LOG = logging.getLogger(__name__)
+
+
+class ExtensionHookCaller(object):
+
+    def __init__(self, config, session):
+        self._url = config.hook_service
+        self._session = session
+
+    def apply(self, kind, obj, app_spec):
+        if self._url is None:
+            return obj
+        url = str(self._url) + "/fiaas/deploy/" + str(kind)
+        dump = json.dumps({"object": obj.as_dict(), "application": self._app_spec_to_dict(app_spec)})
+        response = self._session.post(
+            url,
+            data=dump,
+            headers={'Content-Type': 'application/json', 'Accept': 'application/json'}
+        )
+        if response.status_code == 200:
+            data = response.json()
+            obj.update_from_dict(data)
+        elif response.status_code != 404:
```

Comment:
That is precisely the intention, and 404 seems preferable as it means service-builders don't have to 'actively' ignore kinds they're not interested in, but can simply not implement endpoints that match. At least with most frameworks I know, returning 404 is the default if no route matches, while returning a different error would mean having to implement the route simply to return the error.

I think mapping HTTP statuses to application-level errors is often tricky, but here 404 seems reasonable, as it's the resource-type that's not "found", whereas 501, refers specifically to an HTTP method not being supported at all by the server, for any resource.

Reply:
I see your point, and I agree. It would be bad if we add a new extension point, and an extension starts failing because it doesn't provide a proper status code for that.

On the other hand, when we treat 404 like this, there is no way to notice the difference between these two situations:

- The extension doesn't support this kind
- The configuration is pointing to the wrong path on the extension (Say, setting the `base_url` to `http://service/fiaas/` instead of `http://service/extend/fiaas`)

Probably not a big problem in the short term (or ever), but maybe it should be pointed out as a potential gotcha in the operators guide?

Reply:
Sure, I think the docs should include the details of that 👍